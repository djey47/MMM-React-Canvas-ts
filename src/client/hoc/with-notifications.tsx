import React, { ComponentType, useState } from 'react';

/**
 * Provides a React HOC to provide a gateway between MM2 notifications and components.
 *
 * This HOC relies upon a catcher singleton which has to be plugged into MM2 module lifecycle.
 *
 * @param WrappedComponent component which will be enhanced with notification data
 * @param subscribed list of notification codes to store and relay payload data from
 */
export function withNotifications<P>(
  WrappedComponent: ComponentType<P>,
  subscribed: string[]
): ComponentType<P> {
  const EnhancedComponent: ComponentType<P> = (props: P) => {
    const [data, setData] = useState({});

    NotificationCatcher.getInstance().switchHandler(
      handleNotificationReceived,
      subscribed
    );

    function handleNotificationReceived(notif: string, payload?: unknown) {
      const dataKey = `data_${notif}`;
      setData({
        ...data,
        [dataKey]: payload || {},
      });
    }

    const enhancedPropsWithNotificationData = {
      ...props,
      ...data,
    };
    return <WrappedComponent {...enhancedPropsWithNotificationData} />;
  };

  return EnhancedComponent;
}

export interface NotificationCatcherOptions {
  isDebugMode?: boolean;
}

/**
 * Catcher singleton to be module-wide used
 *
 * It must be instantiated in an early module lifecycle step.
 * Its goal is to relay notifications coming from related lifecycle functions.
 * This will provide a way to plug React components (via HOC) to (socket/non socket) notifications.
 */
export class NotificationCatcher {
  static instance?: NotificationCatcher = undefined;
  notificationHandler: (n: string, p?: unknown) => void;
  subscribedNotifications: string[];
  isInitialized: boolean;
  options: NotificationCatcherOptions;

  constructor(options?: NotificationCatcherOptions) {
    this.notificationHandler = this.defaultHandler;
    this.subscribedNotifications = ['*'];
    this.isInitialized = false;
    this.options = {
      isDebugMode: false,
      ...options,
    };

    Log.log(
      '**** with-notifications: Notification catcher ready for initialization'
    );
  }

  /**
   * @returns the one and only notification catcher instance
   */
  public static getInstance(options?: NotificationCatcherOptions) {
    if (!NotificationCatcher.instance) {
      NotificationCatcher.instance = new NotificationCatcher();
    }

    // Update options if necessary
    NotificationCatcher.instance.options = {
      ...NotificationCatcher.instance.options,
      ...options,
    };

    return NotificationCatcher.instance;
  }

  /**
   * Makes catcher instance process notifications with a custom handler
   * @param handler relay function to be invoked on subscribed notification received
   * @param subscribed list of notification codes to store and relay payload data from ('*' will relay everything)
   */
  public switchHandler(
    handler: (n: string, p?: unknown) => void,
    subscribed: string[]
  ) {
    if (handler == this.notificationHandler) {
      return;
    }

    this.notificationHandler = handler;
    this.subscribedNotifications = subscribed;

    this.debugLog(
      `Notification catcher switched to handler ${handler} for ${subscribed}`
    );
  }

  /**
   * Passes a notification to the catcher if satisfies subscription
   *
   * To be used typically in MM2 lifecycle functions
   *
   * @param notif notification code
   * @param payload notification contents, eventually
   */
  public catchNotification(notif: string, payload?: unknown) {
    if (
      this.subscribedNotifications.includes('*') ||
      this.subscribedNotifications.includes(notif)
    ) {
      this.debugLog(
        `Notification catcher relaying notification {${notif}:${payload}}`
      );

      this.notificationHandler(notif, payload);
    } else {
      this.debugLog(
        `Notification catcher rejecting notification without subscription {${notif}:${payload}}`
      );
    }
  }

  private defaultHandler(n: string, p?: unknown) {
    this.debugLog(
      `Notification catcher processing notification with default handler {${n}:${p}}`
    );
  }

  private debugLog(message: string) {
    if (this.options.isDebugMode) {
      Log.log(`**** with-notifications: ${message}`);
    }
  }
}
