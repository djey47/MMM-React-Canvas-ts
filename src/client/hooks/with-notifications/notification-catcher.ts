export interface NotificationCatcherOptions {
  isDebugMode?: boolean;
}

/**
 * Catcher singleton to be module-wide used
 *
 * It must be instantiated in an early module lifecycle step.
 * Its goal is to relay notifications coming from related lifecycle functions.
 * This will provide a way to plug React components (via HOC) to (socket/non socket) notifications.
 */
export class NotificationCatcher {
  static instance?: NotificationCatcher = undefined;
  notificationHandler: (n: string, p?: unknown) => void;
  subscribedNotifications: string[];
  isInitialized: boolean;
  options: NotificationCatcherOptions;

  constructor(options?: NotificationCatcherOptions) {
    this.notificationHandler = this.defaultHandler;
    this.subscribedNotifications = ['*'];
    this.isInitialized = false;
    this.options = {
      isDebugMode: false,
      ...options,
    };

    Log.log(
      '**** with-notifications: Notification catcher ready for initialization'
    );
  }

  /**
   * @returns the one and only notification catcher instance
   */
  public static getInstance(options?: NotificationCatcherOptions) {
    if (!NotificationCatcher.instance) {
      NotificationCatcher.instance = new NotificationCatcher();
    }

    // Update options if necessary
    NotificationCatcher.instance.options = {
      ...NotificationCatcher.instance.options,
      ...options,
    };

    return NotificationCatcher.instance;
  }

  /**
   * Makes catcher instance process notifications with a custom handler
   * @param handler relay function to be invoked on subscribed notification received
   * @param subscribed list of notification codes to store and relay payload data from ('*' will relay everything)
   */
  public switchHandler(
    handler: (n: string, p?: unknown) => void,
    subscribed: string[]
  ) {
    if (
      handler.toString() === this.notificationHandler.toString() &&
      subscribed === this.subscribedNotifications
    ) {
      return;
    }

    this.notificationHandler = handler;
    this.subscribedNotifications = subscribed;

    this.debugLog(
      `Notification catcher switched to handler ${handler} for ${subscribed}`
    );
  }

  /**
   * Passes a notification to the catcher if satisfies subscription
   *
   * To be used typically in MM2 lifecycle functions
   *
   * @param notif notification code
   * @param payload notification contents, eventually
   */
  public catchNotification(notif: string, payload?: unknown) {
    if (
      this.subscribedNotifications.includes('*') ||
      this.subscribedNotifications.includes(notif)
    ) {
      this.debugLog(
        `Notification catcher relaying notification {${notif}:${payload}}`
      );

      this.notificationHandler(notif, payload);
    } else {
      this.debugLog(
        `Notification catcher rejecting notification without subscription {${notif}:${payload}}`
      );
    }
  }

  private defaultHandler(n: string, p?: unknown) {
    this.debugLog(
      `Notification catcher processing notification with default handler {${n}:${p}}`
    );
  }

  private debugLog(message: string) {
    if (this.options.isDebugMode) {
      Log.log(`**** with-notifications: ${message}`);
    }
  }
}
