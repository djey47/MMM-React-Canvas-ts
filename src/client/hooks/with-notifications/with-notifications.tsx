import { useState } from 'react';
import { NotificationCatcher } from './notification-catcher';

interface WithNotificationsData {
  [key: string]: unknown | object;
}

/**
 * Provides a React hook acting as a gateway between MM2 notifications and components.
 *
 * This hook relies upon a catcher singleton which has to be plugged into MM2 module lifecycle.
 *
 * @param subscribed list of notification codes to store and relay payload data from
 */
const useWithNotifications = (subscribed: string[]): WithNotificationsData => {
  const [data, setData] = useState({});

  NotificationCatcher.getInstance().switchHandler(
    handleNotificationReceived,
    subscribed
  );

  function handleNotificationReceived(notif: string, payload?: unknown) {
    const dataKey = `data_${notif}`;
    setData({
      ...data,
      [dataKey]: payload || {},
    });
  }

  return { ...data };
}

export default useWithNotifications;
