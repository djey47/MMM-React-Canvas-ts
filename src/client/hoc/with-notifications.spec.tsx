import { FunctionComponent } from "react";
import { act, create, ReactTestRenderer } from 'react-test-renderer';
import { withNotifications, NotificationCatcher } from "./with-notifications";

describe('with-notifications HOC', () => {
  describe('withNotifications function', () => {
    it('should enhance props without notification', () => {
      // given-when
      const ActualEnhanced = withNotifications(TestComponent, ['TOPIC']);

      // then
      const tree = create(<ActualEnhanced prop1="val1" />).toJSON();
      expect(tree).toMatchSnapshot();
    });    
    
    it('should enhance props with notification and data', () => {
      // given
      const ActualEnhanced = withNotifications(TestComponent, ['TOPIC']);
      let root: ReactTestRenderer;
      act(() => {
        root = create(<ActualEnhanced prop1="val1" />)
      }); 

      // when
      act(() => {
        NotificationCatcher.getInstance().catchNotification('TOPIC', { data: 1});
        root.update(<ActualEnhanced prop1="val1" prop2="val2" />)
      });

      // then
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(root!.toJSON()).toMatchSnapshot();
    });

    it('should enhance props with notification and no data', () => {
      // given
      const ActualEnhanced = withNotifications(TestComponent, ['TOPIC']);
      let root: ReactTestRenderer;
      act(() => {
        root = create(<ActualEnhanced prop1="val1" />)
      }); 

      // when
      act(() => {
        NotificationCatcher.getInstance().catchNotification('TOPIC');
        root.update(<ActualEnhanced prop1="val1" prop2="val2" />)
      });

      // then
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(root!.toJSON()).toMatchSnapshot();
    });
  });

  describe('NotificationCatcher class', () => {
    describe('constructor', () => {
      it('should init correctly', () => {
        // given-when
        const actual = new NotificationCatcher();

        // then
        expect(actual.isInitialized).toBe(false);
        expect(actual.subscribedNotifications).toEqual(['*']);
        expect(typeof actual.notificationHandler).toBe('function');
      });
    });

    describe('switchHandler method', () => {
      it('should not change subscription if same handler reference provided', () => {
        // given
        const catcher = new NotificationCatcher();
        const defaultHandler = catcher.notificationHandler;

        // when
        catcher.switchHandler(defaultHandler, []);

        // then
        expect(catcher.subscribedNotifications).toEqual(['*']);
      }); 

      it('should change subscription and handler', () => {
        // given
        const catcher = new NotificationCatcher();
        const handler = (n: string, p?: object) => {
          console.log({ n, p });
        };

        // when
        catcher.switchHandler(handler, ['TOPIC']);

        // then
        expect(catcher.notificationHandler).toBe(handler);
        expect(catcher.subscribedNotifications).toEqual(['TOPIC']);
      }); 
    });

    describe('catchNotification method', () => {
      const handlerMock = jest.fn()
 
      beforeEach(() => {
        handlerMock.mockReset();
      });

      it('should not invoke handler if notification does not match subscription', () => {
        // given
        const catcher = new NotificationCatcher();
        catcher.switchHandler(handlerMock, ['NOTIF']);

        // when
        catcher.catchNotification('NOTIF2', {});

        // then
        expect(handlerMock).not.toHaveBeenCalled();
      });      
      
      it('should invoke handler if notification matches subscription', () => {
        // given
        const catcher = new NotificationCatcher();
        catcher.switchHandler(handlerMock, ['NOTIF']);

        // when
        catcher.catchNotification('NOTIF', {});

        // then
        expect(handlerMock).toHaveBeenCalledWith('NOTIF', {});
      });      

      it('should invoke handler if subscription to everything', () => {
        // given
        const catcher = new NotificationCatcher();
        catcher.switchHandler(handlerMock, ['*']);

        // when
        catcher.catchNotification('NOTIF', {});

        // then
        expect(handlerMock).toHaveBeenCalledWith('NOTIF', {});
      });

      it('should invoke default handler', () => {
        // given
        const catcher = new NotificationCatcher();

        // when-then
        catcher.catchNotification('NOTIF', {});
      });
    });

    describe('getInstance static method', () => {
      it('should return existing instance', () => {
        // given-when-then
        expect(NotificationCatcher.getInstance()).toBeDefined();
      });
    });
  });
});

interface TestComponentProps {
  prop1: string;
  prop2?: string;
  data_TOPIC?: object;
}

const TestComponent: FunctionComponent<TestComponentProps> = (props: TestComponentProps) => {
  return (<div {...props}>Test component</div>);
};
