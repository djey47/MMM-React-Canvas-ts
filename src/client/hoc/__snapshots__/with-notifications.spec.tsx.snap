// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with-notifications HOC withNotifications function should enhance props with notification and data 1`] = `
<div
  data_TOPIC={
    {
      "data": 1,
    }
  }
  prop1="val1"
  prop2="val2"
>
  Test component
</div>
`;

exports[`with-notifications HOC withNotifications function should enhance props with notification and no data 1`] = `
<div
  data_TOPIC={{}}
  prop1="val1"
  prop2="val2"
>
  Test component
</div>
`;

exports[`with-notifications HOC withNotifications function should enhance props without notification 1`] = `
<div
  prop1="val1"
>
  Test component
</div>
`;
